hail/src/main/scala/is/hail/utils/FlipbookIterator.scala

```diff
@@ -84,6 +86,65 @@ object FlipbookIterator {
     StagingIterator(sm)
 
   def empty[A] = StagingIterator(StateMachine.terminal[A])
+
+  def multiZipJoin[A: ClassTag](
+    its: IndexedSeq[FlipbookIterator[A]],
+    default: A,
+    ord: (A, A) => Int
+  ): FlipbookIterator[Array[A]] = {
+    val staging = its.map(_.toStagingIterator)
+    val sm = new StateMachine[Array[A]] {
+      private val len = its.length
+      var value: Array[A] = Array.fill(len)(default)
+      var isValid = true
+      var buf = new ArrayBuffer[Int]
+      def advance() {
+        buf.clear()
+        var smallest = -1
+        var i = 0
+        while (i < value.length) {
+          value(i) =  default
+          i += 1
+        }
+
+        i = 0
+        staging.map(_.stage())
+        while (i < its.length) {
```

Comment:
As we discussed, this loop can be O(n) instead of O(N) by using a priority queue.